qiskit/dagcircuit/_dagnode.py

```diff
@@ -0,0 +1,146 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2019, IBM.
+#
+# This source code is licensed under the Apache License, Version 2.0 found in
+# the LICENSE.txt file in the root directory of this source tree.
+
+"""
+   Object to represent the information at a node in the DAGCircuit
+"""
+import copy
+
+
+class DAGNode:
+    """
+    Object to represent the information at a node in the DAGCircuit
+
+    It is used as the return value from *_nodes() functions and can
+    be supplied to functions that take a node.
+
+    """
+    def __init__(self, data_dict, nid=-1):
+        """ Create a node """
+        self._node_id = nid
+        self.data_dict = data_dict
+
+    @property
+    def type(self):
+        """ Returns a str which is the type of the node else None"""
+        return self.data_dict.get('type')
+
+    @property
+    def op(self):
+        """ Returns the Instruction object corresponding to the op for the node else None"""
+        return self.data_dict.get('op')
+
+    @property
+    def name(self):
+        """ Returns a str which is the name of the node else None"""
+        return self.data_dict.get('name')
+
+    @property
+    def qargs(self):
+        """
+        Returns list of (QuantumRegister, int) tuples where the int is the index
+        of the qubit else an empty list
+        """
+        return self.data_dict.get('qargs', [])
```

Comment:
Just a comment that the `data_dict` should not contain `qargs` and `cargs`. This information is already in the edge that connects to this node. So we should read it from there, rather than maintain extra info in our nodes. Can you see if this is easy to remove? otherwise, we can do a followup.

Reply:
I don't think this would be easy to remove as `node.qargs` and `node.cargs` are used frequently, often when iterating over a set of nodes (so not looking at edge data). It could be possible if we created a new method on the DAG that returned the edges as well as the DAGNodes. I agree it would be good to remove this duplication, but I think it should be a separate issue.

Reply:
I think there may be a confusion here. I'm not saying we should remove `node.qargs` and `node.cargs`. We should change the implementation of this getter method so it is not reading it from the data_dict. It should read it from `dag.multi_graph`, which already has this information. Putting this information in the data_dict is redundant and prone to error.

Reply:
actually thinking about this, i think this is more involved and probably left for later. the question is whether nodes should know about where they are attached. I think not, but i"ll open a follow up